### ISSUE TITLE
Support new data type: `TimestampWithTimeZone`

### ISSUE THREAD

**Issue by xiangfu0** [apache/pinot#13004]
This is also a long time ask, similar to presto type: [TIMESTAMP WITH TIME ZONE](https://prestodb.io/docs/current/language/types.html#timestamp-with-time-zone)

```
Instant in time that includes the date and time of day with a time zone. Values of this type are rendered using the time zone from the value.

Example: TIMESTAMP '2001-08-22 03:04:05.321 America/Los_Angeles'
```

**Comment by gortiz** [comment]
Just as a reminder, this type in Postgres does not actually include timezone. Instead, at disk level, this type is just a normal timestamp. As explained by [the doc](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-INPUT-TIME-STAMPS):

> For timestamp with time zone, the internally stored value is always in UTC (Universal Coordinated Time, traditionally known as Greenwich Mean Time, GMT). An input value that has an explicit time zone specified is converted to UTC using the appropriate offset for that time zone. If no time zone is stated in the input string, then it is assumed to be in the time zone indicated by the system's [TimeZone](https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-TIMEZONE) parameter, and is converted to UTC using the offset for the timezone zone.

> When a timestamp with time zone value is output, it is always converted from UTC to the current timezone zone, and displayed as local time in that zone. To see the time in another time zone, either change timezone or use the AT TIME ZONE construct (see [Section 9.9.4](https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-ZONECONVERT)).

> Conversions between timestamp without time zone and timestamp with time zone normally assume that the timestamp without time zone value should be taken or given as timezone local time. A different time zone can be specified for the conversion using AT TIME ZONE.

So basically the timezone feature is mostly a frontend property. When timezone is included in the type, values are converted to UTC. When timezone is not included, the received timezone is considered to be based on the server timezone.

**Comment by hpvd** [comment]
`Example: TIMESTAMP '2001-08-22 03:04:05.321 America/Los_Angeles'`

I'm a little unsure if this format may not lead to confusion.
The problem I see when using these values is, that Dayllight saving time (DST) / summer and winter time is not reflected in this.
Since summer and winter time is not aligned all over the world,
see https://en.wikipedia.org/wiki/Daylight_saving_time_by_country
some calculations on base of these values may get complicate/prone to errors.

One approach to solve this may be adding `UTC+X` to timestamp instead

Example:
TIMESTAMP '2001-07-23 03:04:05.321 UTC+2' -> Germany/Berlin/summertime
TIMESTAMP '2001-12-24 03:04:05.321 UTC+1' -> Germany/Berlin/wintertime

**Comment by gortiz** [comment]
Format should be ISO-8601 (which is the one used in JSR-310).

**Comment by hpvd** [comment]
btw the "official" letter "T" as a delimiter between Date and Time in this example, makes it also easier to read/use
`2007-04-05T12:30−02:00`
https://en.wikipedia.org/wiki/ISO_8601#Combined_date_and_time_representations

**Comment by hpvd** [comment]
+1 on this feature

**Comment by Jackie-Jiang** [comment]
@mghildiy It should be similar to how the `TIMESTAMP` data type is handled (you may search the usage of `DataType.TIMESTAMP`). We need to design the internal storage format (presto uses long format and we may use the same way to store it), and how it interacts with other `TIMESTAMP WITH TIME ZONE` and `TIMESTAMP`

**Comment by mghildiy** [comment]
@Jackie-Jiang , thanks for the lead.

**Comment by mghildiy** [comment]
Which part of codebase takes care of internal storage format?

**Comment by Jackie-Jiang** [comment]
Each `DataType` has a `getStoredType()`.
`BOOLEAN` is stored as `INT`, `TIMESTAMP` is stored as `LONG`, `JSON` is stored as `STRING`. Take a look at its usage

**Comment by mghildiy** [comment]
Hi, I may take some time here due to an unplanned change in events.

**Comment by mghildiy** [comment]
I am restarting work on it.

**Comment by mghildiy** [comment]
I have a query here.
This data type needs to be a composite one as its combination of instant value and time zone.
Hnece, Internal storage format would be a combination of long(for instant value in UTC) and string/int(for timezone). Am I correct?

**Comment by gortiz** [comment]
> This data type needs to be a composite one as its combination of instant value and time zone.
Heece, internal storage format would be a combination of long(for instant value in UTC) and string/int(for timezone). Am I correct?

I don't think that should be the case. I think what Postgres does is pretty smart here. Basically at storage level, `timestamp (without time zone)` and `timestamp with time zone` are equivalent. They just store millis since epoch (which is 1/1/1970 00:00 UTC). The only difference is how that is interpreted when written/reading. When `timestamp without time zone` is used, all writes/reads are interpreted to be UTC (or the timezone used by the server, I guess). When `timestamp with time zone` is used, then it follows a more complex logic that depends on the timezone of the client/connection.

I strongly recommend to read [postgres documentation page describing datetime datatypes](https://www.postgresql.org/docs/current/datatype-datetime.html), this [stackoverflow answer](https://stackoverflow.com/a/5876276) and specially [this post from yugabyte](https://www.yugabyte.com/blog/postgresql-timestamps-timezones/)

**Comment by gortiz** [comment]
Exactly that is the case and that is what we would like to have.

**Comment by mghildiy** [comment]
> Exactly that is the case and that is what we would like to have.

I have a query here then. So when query arrives at database engine, does it also contain client's timezone information?

**Comment by gortiz** [comment]
I don't think so, and therefore I'm not sure if this issue is well defined. I think what we need is to add `TIMESTAMP WITH TIMEZONE` as a runtime type, but it should be stored as a `TIMESTAMP` (and therefore a long), converted at UTC.

What we need to add then is the ability to write literals as `'2001-08-22 03:04:05.321 America/Los_Angeles'`, which would be of type `TIMESTAMP WITH TIMEZONE`. Then we need functions to transform from one timezone to another. Probably some of the timestamp functions we have right now support timezone and we will need to think about what to do with them.

**Comment by mghildiy** [comment]
Thanks  @gortiz.
Any more inputs by anyone else?

**Comment by gortiz** [comment]
I've found [this document](https://facebookincubator.github.io/velox/develop/timestamp.html) describing TIMESTAMP and TIMESTAMP WITH TIMEZONE in Velox and I think it we can get use of this. Especially the concept and physical representation is similar to what we want to achieve here

**Comment by mghildiy** [comment]
> I've found [this document](https://facebookincubator.github.io/velox/develop/timestamp.html) describing TIMESTAMP and TIMESTAMP WITH TIMEZONE in Velox and I think it we can get use of this. Especially the concept and physical representation is similar to what we want to achieve here

Thanks @gortiz for the link. Looking into it.

**Comment by mghildiy** [comment]
After going through that document, my approach would be:

- Represent timestamp with time zone similar to timestamp ie a long value, with a part of bits used to store timezone.
- User's tiemzone input(a string) is mapped to an integer representing one of the tiemzones and viceversa.
- A mapper/converter class takes care of mapping mentioned above.

**Comment by vrajat** [comment]
Another comprehensive [blog|https://kaiwern.com/posts/2021/07/20/what-you-need-to-know-about-postgresql-timezone/] on how time zone is handled in Postgres. +1 to storing timestamp in UTC at the storage level. A couple of other configs required are timezone of the server and client. The timezone of the server can be a config. Timezone of the client can be set through a query option ? 

With these, it should be possible to provide all the functionality similar to PG.

**Comment by mghildiy** [comment]
I have a query. User input would be something like '2021-01-01 01:01:01.001 America/Chicago', and not with offsets. Am I right?

**Comment by mghildiy** [comment]
[This](https://github.com/apache/pinot/pull/14526) is the PR I have created. It's a WIP currently.

**Comment by Jackie-Jiang** [comment]
We've got designs of different flavors:
- TIMESTAMP NO TIME ZONE: #14398
- TIMESTAMP WITH TIME ZONE: #14526 

Here is the PostgreSQL [documentation](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME) for reference. The internal storage format should be long value (millis since epoch), and we need to align on the input value parsing and output result format for these data types:
- What input format is allowed
- If input doesn't have time zone info, which time zone to use to parse the value
- What output format to use

cc @xiangfu0 @gortiz
